noun/retrieve.c

```diff
@@ -1770,113 +1518,122 @@ u3_mur_words(const c3_w* key_w, c3_w len_w)
   }
 }
 
-/* u3_mur_both():
+/* u3r_mug_bytes(): Compute the mug of `buf`, `len`, LSW first.
 **
-**   Join two murs.
+**   XX temporary, reconcile with u3r_mug_words
+*/
+c3_w
+u3r_mug_bytes(const c3_y *buf_y,
+              c3_w        len_w)
+{
+  c3_w syd_w = 0xcafebabe;
+
+  while ( 1 ) {
+    c3_w haz_w, ham_w;
+
+    MurmurHash3_x86_32(buf_y, len_w, syd_w, &haz_w);
+    ham_w = (haz_w >> 31) ^ (haz_w & 0x7fffffff);
+
+    if ( 0 != ham_w ) return ham_w;
+    else syd_w++;
+  }
+}
+
+/* u3r_mug_d(): Compute the mug of `num`, LSW first.
+*/
+c3_w
+u3r_mug_d(c3_d num_d)
+{
+  c3_w buf_w[2];
+
+  buf_w[0] = (c3_w)(num_d & 0xffffffffULL);
+  buf_w[1] = (c3_w)(num_d >> 32ULL);
+
+  return u3r_mug_words(buf_w, 2);
+}
+
+/* u3r_mug_both(): Join two mugs.
 */
 c3_w
-u3_mur_both(c3_w lef_w, c3_w rit_w)
+u3r_mug_both(c3_w lef_w, c3_w rit_w)
 {
   c3_w ham_w = lef_w ^ (0x7fffffff ^ rit_w);
```

Comment:
Why do we trim the top bit off only rit_w, and not lef_w? In fact, why do we trim at all? Aren't they both 31-bit values to begin with?

Reply:
We're not trimming here, just inverting the bits of `rit_w` before XOR with `lef_w`. As for why, idk. It seems plausible that this would increase the variance of combined values. But that's just a guess.

Reply:
Oh, yeah, I read that wrong. Actually, I think it's there to make it less likely to get zeros. If both sides of the pair are identical, a straight XOR of the sides would yield 0, which would then collide more. Seems legit